### PR TITLE
Generate a separate test function for every header

### PR DESCRIPTION
**Pull request by jdub**
- Header tests can now be run selectively, e.g.
  `cargo test header` will test all headers
  `cargo test union` will test headers with 'union' in the file name
- The list of test functions is generated by `build.rs`, so never needs to be updated
- Clever approach suggested by @fitzgen 

This PR depends on #204, so will include those changes until they're merged into master and this can be rebased. But I'll keep the commits separate.

### CODE REVIEW

File: libbindgen/build.rs [hunk 1]
```diff
@@ -14,6 +13,42 @@ mod codegen {
...
+            .map(|result| result.expect("Couldn't read header file"));
+
+        for entry in entries {
+            match entry.path().extension().and_then(OsStr::to_str) {
+                Some("h") | Some("hpp") => {
+                    let func = entry.file_name().to_str().unwrap()
+                        .replace(|c| !char::is_alphanumeric(c), "_")
+                        .replace("__", "_")
+                        .to_lowercase();
+                    let _ = writeln!(dst, "test_header!(header_{}, {:?});",
```
Comment (emilio) [inline libbindgen/build.rs]: I *think* You need to write these in a block to avoid https://github.com/rust-lang/rust/issues/35560

Comment (jdub) [inline libbindgen/build.rs]: The `include!()` is not itself in a function, so it just pulls in a bunch of test function definitions, and doesn't interact with that bug. I'm getting the same pass/fail rate as #204, only pretty. 😄

```
test result: FAILED. 120 passed; 1 failed; 0 ignored; 0 measured
```

Comment (emilio) [inline libbindgen/build.rs]: Ok, sounds good then :)

Do you plan to land everything here (and thus close #204), or splitting the last commit?

**Comment by emilio** [comment]
In any case, just update the test for now, let's get this landed, and then polish it later :)

File: libbindgen/build.rs (hunk 1 above)
Comment (jdub) [inline libbindgen/build.rs]: I'll rebase this once #204 has landed -- further comments over there. :-)

**Comment by fitzgen** [comment]
Thanks for doing this one too :-D

**Pull request merged**